**Importing Important Libraries**

In [4]:
import numpy as np 
import pandas as pd 

**Read Our Data From file and 
  storing in movies and credits**

In [5]:
movies=pd.read_csv('tmdb_5000_movies.csv')
credits=pd.read_csv('tmdb_5000_credits.csv')

In [6]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [7]:
movies = movies.merge(credits, on='title')

**Keeps only Important Column For recomendation**


In [8]:
#genres , id , Keywords , title , overview , cast , crew 
movies = movies[['genres','id','keywords','overview','title','cast','crew']]

In [9]:
movies.head(2)

genres     id  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  19995   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...    285   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [10]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   genres    4809 non-null   object
 1   id        4809 non-null   int64 
 2   keywords  4809 non-null   object
 3   overview  4806 non-null   object
 4   title     4809 non-null   object
 5   cast      4809 non-null   object
 6   crew      4809 non-null   object
dtypes: int64(1), object(6)
memory usage: 263.1+ KB


**Checking the null values in Movies data**

In [11]:
movies.isnull().sum()

genres      0
id          0
keywords    0
overview    3
title       0
cast        0
crew        0
dtype: int64

**We are Droping the missinig value of overview from Movies dataset**

In [12]:
movies.dropna(inplace=True)

In [13]:
movies.iloc[0].genres


'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

**Converting the wierd data in meaningfull data**

Wierd Data ='[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

**Into** 


**Action , Adventure , Fantasy , Sci-Fi** 



In [14]:
import ast
def convert (obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['name'])
        return L

In [15]:
movies['genres'] = movies['genres'].apply(convert)

In [16]:
movies.head()

genres      id                                           keywords  \
0     [Action]   19995  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [Adventure]     285  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2     [Action]  206647  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3     [Action]   49026  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4     [Action]   49529  [{"id": 818, "name": "based on novel"}, {"id":...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

**Same thing we do with keywords**

In [17]:
movies['keywords']= movies['keywords'].apply(convert)

In [18]:
movies.head()

genres      id          keywords  \
0     [Action]   19995   [culture clash]   
1  [Adventure]     285           [ocean]   
2     [Action]  206647             [spy]   
3     [Action]   49026       [dc comics]   
4     [Action]   49529  [based on novel]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [19]:
import ast

def convert3(text):
    L = []
    counter = 0
    
    # Check if the input is already a list
    if isinstance(text, str):  # If it's a string, use ast.literal_eval to parse it
        text = ast.literal_eval(text)
    
    # Iterate over the list and extract up to 3 names
    for i in text:
        if counter < 3:
            if isinstance(i, dict) and 'name' in i:  # Handle dictionary elements
                L.append(i['name'])
            elif isinstance(i, str):  # Handle string elements
                L.append(i)
            counter += 1
    
    return L

movies['cast'] = movies['cast'].apply(convert3)




In [20]:
movies['cast'] = movies['cast'].apply(convert3)

In [21]:
movies.head(4)

genres      id         keywords  \
0     [Action]   19995  [culture clash]   
1  [Adventure]     285          [ocean]   
2     [Action]  206647            [spy]   
3     [Action]   49026      [dc comics]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...

In [22]:
def fetch_director(text):
    L = []
    for i in ast.literal_eval(text):
        if i['job'] == 'Director':
            L.append(i['name'])
    return L 

In [23]:
movies['crew'] = movies['crew'].apply(fetch_director)

In [24]:
movies.head()

genres      id          keywords  \
0     [Action]   19995   [culture clash]   
1  [Adventure]     285           [ocean]   
2     [Action]  206647             [spy]   
3     [Action]   49026       [dc comics]   
4     [Action]   49529  [based on novel]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

**Removing the spaces of columns data**

In [25]:
def collapse(L):
    if L:  # Check if L is not None or empty
        L1 = []
        for i in L:
            L1.append(i.replace(" ", ""))
        return L1
    return []  


In [26]:
movies['cast'] = movies['cast'].apply(collapse)
movies['crew'] = movies['crew'].apply(collapse)
movies['genres'] = movies['genres'].apply(collapse)
movies['keywords'] = movies['keywords'].apply(collapse)


In [27]:
movies.head()

genres      id        keywords  \
0     [Action]   19995  [cultureclash]   
1  [Adventure]     285         [ocean]   
2     [Action]  206647           [spy]   
3     [Action]   49026      [dccomics]   
4     [Action]   49529  [basedonnovel]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

In [28]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

**tags = overview + generes + keywords + cast + crew**

In [29]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [30]:
new_df = movies.drop(columns=['overview','genres','keywords','cast','crew'])

In [31]:
new_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

**Coverting list data into String**

In [32]:
new_df['tags'] = new_df['tags'].apply(lambda x:" ".join(x))

**Converting Tags columns data in Lowercase**

In [33]:
new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())

**applying Count vectorization**

In [34]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000,stop_words='english')

In [35]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [36]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [37]:
from sklearn.metrics.pairwise import cosine_similarity

In [38]:
similarity = cosine_similarity(vectors)

**Creating a Function that give five related movies**

In [39]:
def recommend(movie):
    index = new_df[new_df['title'] == movie].index[0]
    distances = sorted(list(enumerate(similarity[index])),reverse=True,key = lambda x: x[1])
    
    for i in distances[1:6]:
        print(new_df.iloc[i[0]].title)
        

In [40]:
recommend('Batman Begins')

The Dark Knight Rises
Raising Helen
The Dark Knight
Batman Forever
Synecdoche, New York


In [43]:
import pickle

In [44]:
pickle.dump(new_df.to_dict() , open('movie_dict.pkl','wb'))

In [45]:
pickle.dump(similarity,open('similarity.pkl','wb'))